File: src/main/scala/io/github/configur8/Property.scala
package io.github.configur8

import io.github.configur8.ExposeMode.Public

/**
 * A typed property which can be serialized and deserialized to a string value
 */
case class Property[T] private(name: String, deserialize: (String => T), serialize: (T => String), exposeMode: ExposeMode) {
  override def toString = name
}

object Property {
  /**
   * Create a Property
   * @param name of the property
   * @param deserialize function
   * @param serialize function
   * @param expose mode which determines if the value should be publicly accessible or concealed
   * @return The property
   */
  def of[T](name: String, deserialize: String => T, serialize: (T => String) = (i: T) => i.toString, exposeMode: ExposeMode = Public): Property[T] = Property[T](name, deserialize, serialize, exposeMode)

  /**
   * Pre-configured Char Property
   */
  def char(name: String, exposeMode: ExposeMode = Public): Property[Char] = Property.of(name, (i: String) => i.charAt(0), exposeMode = exposeMode)

  /**
   * Pre-configured String Property
   */
  def string(name: String, exposeMode: ExposeMode = Public): Property[String] = Property.of(name, (i: String) => i, exposeMode = exposeMode)

  /**
   * Pre-configured Integer Property
   */
  def integer(name: String, exposeMode: ExposeMode = Public): Property[Int] = Property.of(name, (i: String) => i.toInt, exposeMode = exposeMode)

  /**
   * Pre-configured Long Property
   */
  def long(name: String, exposeMode: ExposeMode = Public): Property[Long] = Property.of(name, (i: String) => i.toLong, exposeMode = exposeMode)

  /**
   * Pre-configured boolean Property
   */
  def boolean(name: String, exposeMode: ExposeMode = Public): Property[Boolean] = Property.of(name, (i: String) => i.toBoolean, exposeMode = exposeMode)
}

File: java/src/main/java/io/github/configur8/Property.java
package io.github.configur8;

/**
 * A typed property which can be serialized and deserialized to a string value
 */
public class Property<T> {

    public final String name;
    public final Deserializer<T> deserialize;
    public final Serializer<T> serialize;
    public final ExposeMode exposeMode;

    private Property(String name, Deserializer<T> deserialize, Serializer<T> serialize, ExposeMode exposeMode) {
        this.name = name;
        this.deserialize = deserialize;
        this.serialize = serialize;
        this.exposeMode = exposeMode;
    }

    @Override
    public String toString() {
        return name;
    }

    /**
     * Create a Property
     * @param name of the property
     * @param deserialize function
     * @param serialize function
     * @return The property
     */
    public static <T> Property<T> of(String name, Deserializer<T> deserialize, Serializer<T> serialize) {
        return of(name, deserialize, serialize, ExposeMode.Public);
    }

    /**
     * Create a Property
     * @param name of the property
     * @param deserialize function
     * @return The property
     */
    public static <T> Property<T> of(String name, Deserializer<T> deserialize) {
        return of(name, deserialize, Object::toString, ExposeMode.Public);
    }

    /**
     * Create a Property
     * @param name of the property
     * @param deserialize function
     * @param expose mode which determines if the value should be publicly accessible or concealed
     * @return The property
     */
    public static <T> Property<T> of(String name, Deserializer<T> deserialize, ExposeMode exposeMode) {
        return of(name, deserialize, Object::toString, exposeMode);
    }


    /**
     * Create a Property
     * @param name of the property
     * @param deserialize function
     * @param serialize function
     * @param expose mode which determines if the value should be publicly accessible or concealed
     * @return The property
     */
    public static <T> Property<T> of(String name, Deserializer<T> deserialize, Serializer<T> serialize, ExposeMode exposeMode) {
        return new Property<>(name, deserialize, serialize, exposeMode);
    }

    /**
     * Pre-configured String Property
     */
    public static Property<String> string(String name) {
        return string(name, ExposeMode.Public);
    }

    /**
     * Pre-configured String Property
     */
    public static Property<String> string(String name, ExposeMode exposeMode) {
        return new Property<>(name, String::new, s -> s, exposeMode);
    }

    /**
     * Pre-configured Int Property
     */
    public static Property<Integer> integer(String name) {
        return integer(name, ExposeMode.Public);
    }

    /**
     * Pre-configured Int Property
     */
    public static Property<Integer> integer(String name, ExposeMode exposeMode) {
        return new Property<>(name, Integer::parseInt, Object::toString, exposeMode);
    }

    /**
     * Pre-configured Long Property
     */
    public static Property<Long> aLong(String name) {
        return aLong(name, ExposeMode.Public);
    }

    /**
     * Pre-configured Long Property
     */
    public static Property<Long> aLong(String name, ExposeMode exposeMode) {
        return new Property<>(name, Long::parseLong, Object::toString, exposeMode);
    }

    /**
     * Pre-configured Boolean Property
     */
    public static Property<Boolean> bool(String name) {
        return bool(name, ExposeMode.Public);
    }

    /**
     * Pre-configured Boolean Property
     */
    public static Property<Boolean> bool(String name, ExposeMode exposeMode) {
        return new Property<>(name, Boolean::parseBoolean, Object::toString, exposeMode);
    }

    /**
     * Pre-configured Char Property
     */
    public static Property<Character> character(String name) {
        return character(name, ExposeMode.Public);
    }

    /**
     * Pre-configured Char Property
     */
    public static Property<Character> character(String name, ExposeMode exposeMode) {
        return new Property<>(name, (in) -> in.charAt(0), Object::toString, exposeMode);
    }
}
